[FLINK-13904][checkpointing] Avoid competition of checkpoint triggering

ifndef-SleePy opened the pull request:
## What is the purpose of the change

* This is a part of reworking threading model of `CheckpointCoordinator`
* This PR tries to avoid competition of checkpoint/savepoint triggering

## Brief change log

* Make trigger thread of `CheckpointCoordinator` single-threaded
  * This is a preparation of running triggering in main thread executor
* Get rid of trigger lock of `CheckpointCoordinator`
* Manually schedule periodic checkpoint trigger instead of `scheduleAtFixedRate`
  * It simplifies the trigger logic
  * It reduces a lot of cancellations of periodic triggering
  * This is also a preparation of running triggering in main thread, because currently `scheduleAtFixedRate` is not supported yet in main thread executor
* It also reworks some test case structures to satisfy code style checking

## Verifying this change

* This change is already covered by existing tests
* Actually in this PR, there is no new feature introduced. So we just need to make sure all existing tests could pass

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

ifndef-SleePy commented:
@flinkbot run travis

ifndef-SleePy commented:
@flinkbot attention @pnowojski

ifndef-SleePy commented:
Oops! I found a race condition of the last commit of this PR when I implement the follow-up PR.
I didn't find an easy way to solve this for now. I think I could correct it tomorrow.
@pnowojski

pnowojski commented:
Could you include the fix for this bug as a separate fixup commit (I'm in the middle of reviewing of the current version)?

ifndef-SleePy commented:
Hi @pnowojski , I have updated the PR with a hotfix appended.

The race condition scenario is when there are too many in-flight checkpoints, the error handling should not release the lock and check whether we should reschedule trigger or not. Because the `triggerRequestQueued` could be changed by IO thread at this time, if we release the lock.

BTW, some codes of this PR might not be elegant somewhat. There will be some refactoring in the later PRs.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTriggeringTest.java:
```diff
@@ -289,29 +292,29 @@ public void testStopPeriodicScheduler() throws Exception {
-		} catch (CheckpointException e) {
-			assertEquals(CheckpointFailureReason.PERIODIC_SCHEDULER_SHUTDOWN, e.getCheckpointFailureReason());
+		} catch (ExecutionException e) {
+			assertTrue(e.getCause() instanceof CheckpointException);
```
`ExceptionUtils#findThrowable(java.lang.Throwable, java.lang.Class<T>)`?

ifndef-SleePy replied:
Good point!

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTriggeringTest.java:
```diff
@@ -289,29 +292,29 @@ public void testStopPeriodicScheduler() throws Exception {
+			null,
+			false,
+			false);
+		assertFalse(nonPeriodicTriggerResult.isCompletedExceptionally());
```
ditto about waiting for the future to complete? 
```
assertFalse(nonPeriodicTriggerResult.isCompletedExceptionally());
```
 ->
```
nonPeriodicTriggerResult.get();
```
?

ifndef-SleePy replied:
Ditto.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -1306,10 +1312,8 @@ private long getRandomInitDelay() {
-			new ScheduledTrigger(),
-			initDelay, baseInterval, TimeUnit.MILLISECONDS);
+	private ScheduledFuture<?> scheduleTriggerWithDelay(long delay) {
+		return timer.schedule(new ScheduledTrigger(), delay, TimeUnit.MILLISECONDS);
```
What is the benefit/purpose of this change and this commit (`Manually schedule periodic checkpoint trigger instead of scheduleAtFixedRate`) as a whole?

ifndef-SleePy replied:
That's a good question. The direct reason is that currently the main thread executor does not support periodic scheduling, it's not implemented yet. So there are two options.
1. Support periodic scheduling in main thread executor.
2. Manually triggering checkpoint in `CheckpointCoordinator`.

There is a short discussion of the option 1, see https://issues.apache.org/jira/browse/FLINK-13848.

I choose option 2 currently. Because I'm a not fan of the way of periodic scheduling currently. There are a lot of cancellation of periodic scheduling. Cancellation is not accurate and reliable somewhat.

But to be honest, there is a short-coming of option 2. We have to make sure the next trigger will be scheduled under all scenarios, successful, failed, request queued (it's messy of handling queued request now). It's a bit complicated than I expected.

Do you have any suggestion of this?

ifndef-SleePy commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorMasterHooksTest.java:
```diff
@@ -201,7 +201,7 @@ public void testHooksAreCalledOnTrigger() throws Exception {
 
 		// trigger a checkpoint
-		assertTrue(cc.triggerCheckpoint(System.currentTimeMillis(), false));
+		assertFalse(cc.triggerCheckpoint(System.currentTimeMillis(), false).isCompletedExceptionally());
```
Ah, sorry that's my fault. I forgot updating the java doc. I changed the behavior of `triggerCheckpoint` a bit. The return value, `CompletableFuture<CompletedCheckpoint>`, will not complete until this checkpoint is finished. It will not complete immediately after `triggerCheckpoint`. Unless there is an error occurred. It's similar to `triggerSavepoint`.
I did this because eventually the `triggerCheckpoint` will be non-blocking. It will be separated into several stages, pre-checking, initializing the id and checkpointStorageLocation, triggering snapshotting master state, and then triggering snapshotting task state. We can't judge the triggering is successful or not by a boolean return value as before.

So in this case, we can't wait for the future to complete, because it will never complete.

ifndef-SleePy replied:
If there is an error occurred immediately, the future will completed with an exception.
If the triggering is successful, we could check whether the tasks receive the `triggerCheckpoint` message or not.

ifndef-SleePy replied:
I almost missed this message.
Yes, I agree, testing case through mockito is not easy to maintain. It could be a follow-up issue (there are already some in my list). I could come back to this after the main part merged.

ifndef-SleePy commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java:
```diff
@@ -1347,6 +1356,7 @@ public void testTriggerAndConfirmSimpleSavepoint() throws Exception {
 		CompletableFuture<CompletedCheckpoint> savepointFuture = coord.triggerSavepoint(timestamp, savepointDir);
 		assertFalse(savepointFuture.isDone());
 
+		checkpointTriggeredLatch1.await();
```
As mentioned before, the latch is to check if the triggering is successful.

ifndef-SleePy commented on flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/utils/SimpleAckingTaskManagerGateway.java:
```diff
@@ -70,6 +73,10 @@ public void setReleasePartitionsConsumer(BiConsumer<JobID, Collection<ResultPart
 		this.releasePartitionsConsumer = releasePartitionsConsumer;
 	}
 
+	public void setCheckpointConsumer(Consumer<Tuple6<ExecutionAttemptID, JobID, Long, Long, CheckpointOptions, Boolean>> checkpointConsumer) {
```
Ditto, it's used to check if the tasks receive `triggerCheckpoint` message.

ifndef-SleePy commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -535,149 +528,142 @@ public boolean isShutdown() {
-		// this is not done with the coordinator-wide lock, because the 'checkpointIdCounter'
-		// may issue blocking operations. Using a different lock than the coordinator-wide lock,
-		// we avoid blocking the processing of 'acknowledge/decline' messages during that time.
-		synchronized (triggerLock) {
```
The `triggerSavepointInternal` will not invoke `triggerCheckpoint` directly anymore. In the prior commit of this one, I have changed the behavior of `triggerSavepointInternal`, it executes the trigger message in timer thread. So all trigger operations (checkpoint or savepoint) are executed in timer thread (will be replaced by main thread eventually). There will be no competition under this `triggerLock`. I will indicate it in commit message.

ifndef-SleePy replied:
Done.

ifndef-SleePy commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -535,149 +528,142 @@ public boolean isShutdown() {
 
```
I fully understand...
The reason I didn't rework this part is the `triggerCheckpoint` will be changed in later PRs. I was planning to make the non-functional reworking in the very end PRs.

ifndef-SleePy commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -153,8 +155,8 @@
 
-	/** The timestamp (via {@link System#nanoTime()}) when the last checkpoint completed. */
-	private long lastCheckpointCompletionNanos;
+	/** The relative timestamp when the next checkpoint could be triggered. */
```
Actually it's not the time which next checkpoint "should" be triggered. The `earliestRelativeTimeNextCheckpointBeTriggered` means next trigger couldn't be launched until this time due to the restriction of `minPauseBetweenCheckpoints`.

ifndef-SleePy commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -153,8 +155,8 @@
-	/** The timestamp (via {@link System#nanoTime()}) when the last checkpoint completed. */
-	private long lastCheckpointCompletionNanos;
+	/** The relative timestamp when the next checkpoint could be triggered. */
+	private long earliestRelativeTimeNextCheckpointBeTriggered;
```
As mentioned before, "earliest" means it's the earliest time next trigger could be triggered.
"relative" means it's a relative time (`System.nanotime()`).
Maybe `earliestNextCheckpointTriggerRelativeTime`?

pnowojski replied:
I would shorten it to `nextCheckpointTriggerRelativeTime`

ifndef-SleePy replied:
I think the problem is this field includes too much meaning. So it's not easy to give it a short and clear name. I reverted this field back to `lastCheckpointCompletionRelativeTime`. Calculates the `nextCheckpointTriggerRelativeTime` in `checkMinPauseBetweenCheckpoints`, it's much easier to understand with context.

ifndef-SleePy commented:
Hi @pnowojski , thanks for feedback.

I have updated the PR to address review comments, except the commit "Manually schedule periodic checkpoint trigger instead of scheduleAtFixedRate" and "[hotfix] Fix the race condition..." relevant. 

I'm wondering if it worths having or not. As you pointed out, it feels like no so much benefit as expected. It brings some troubles and challenges in this and later PR. Maybe keeping it in a "scheduleAtFixedRate" way as the current is a better choice. It's not perfect but it works and causes less trouble. 

I tend to remove the these two commits (manually schedule periodic trigger relevant) from this PR at the moment. What do you think?

ifndef-SleePy commented:
Hi @pnowojski  again.

I just reverted the commit "Manually schedule periodic checkpoint trigger instead of scheduleAtFixedRate" and hotfix commits on it. After struggling some with it, I think I should keep the current implementation of periodic triggering unless there is a strong reason to change it.

Actually the target of this subtask "Avoid competition between different rounds of checkpoint triggering" is satisfied without the reverted commits.

BTW, the commit message of "addressing review comments" commit is started with "[hotfix] Address review comments". I placed them in the correct position of commit history. They should be squashed before merging.

ifndef-SleePy commented:
@flinkbot run travis

ifndef-SleePy commented on flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/utils/SimpleAckingTaskManagerGateway.java:
```diff
@@ -54,6 +55,8 @@
 
 	private BiConsumer<JobID, Collection<ResultPartitionID>> releasePartitionsConsumer = (ignore1, ignore2) -> { };
 
+	private Consumer<Tuple6<ExecutionAttemptID, JobID, Long, Long, CheckpointOptions, Boolean>> checkpointConsumer = ignore -> { };
```
Good suggestion.

ifndef-SleePy commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java:
```diff
@@ -1466,11 +1477,24 @@ public void testSavepointsAreNotSubsumed() throws Exception {
+					add(checkpointTriggeredLatch2);
+					add(checkpointTriggeredLatch3);
+					add(checkpointTriggeredLatch4);
+					add(checkpointTriggeredLatch5); }});
```
That's really a nice suggestion!
I thought there might be some problems with a manually executor for these cases. But after did a testing, I realized all cases work well with manually executor. Moreover in this way, some time based case could be simplified a lot.
BTW, I keep some "dead code" for now, for example, `TestingScheduledExecutor`, and the changes of `SimpleAckingTaskManagerGateway`. It might be useful for the later PR or others.

pnowojski replied:
I'm glad that it worked out :)

> BTW, I keep some "dead code" for now, for example, TestingScheduledExecutor, and the changes of SimpleAckingTaskManagerGateway. It might be useful for the later PR or others.

I think I would revisit this at the end of this effort.

ifndef-SleePy commented:
Hi @pnowojski , thanks for reviewing! I have addressed the comments. Could you take a look?

ifndef-SleePy commented on flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/utils/SimpleAckingTaskManagerGateway.java:
```diff
@@ -55,7 +54,7 @@
 	private BiConsumer<JobID, Collection<ResultPartitionID>> releasePartitionsConsumer = (ignore1, ignore2) -> { };
 
-	private Consumer<Tuple6<ExecutionAttemptID, JobID, Long, Long, CheckpointOptions, Boolean>> checkpointConsumer = ignore -> { };
+	private CheckpointConsumer checkpointConsumer = null;
```
Done

ifndef-SleePy commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java:
```diff
@@ -316,6 +319,9 @@
 	private CheckpointCoordinator checkpointCoordinator;
 
+	/** TODO, replace it with main thread executor. */
+	private ScheduledExecutorService checkpointCoordinatorTimer;
```
Done

ifndef-SleePy commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java:
```diff
@@ -593,6 +599,10 @@ public void failJobDueToTaskFailure(Throwable cause, ExecutionAttemptID failingT
 			}
 		);
 
+		checkpointCoordinatorTimer = Executors.newSingleThreadScheduledExecutor(
```
Done

ifndef-SleePy commented:
Hi @pnowojski , just addressed the comments :)

ifndef-SleePy commented:
BTW, I just found an interesting thing which I never met before. The order of commit history on this Github PR (at-least in my web) is not same with my private branch which the PR is based on (I did a `git rebase -i ...` to make the `squash` easier). I'm not sure if the problem still exists or not when merging the PR. 
Maybe I should do the `squash` myself before merging.

pnowojski commented:
> The order of commit history on this Github PR (at-least in my Chrome) is not same with my private branch 

I will take care of it while merging. Welcome to the elite club of people affected/annoyed by:

https://github.com/isaacs/github/issues/386

please upvote it so maybe finally github will fix this. Github orders the commit by theirs modification timestamp, so until this is fixed, the dirty solution is to do:
```
git rebase -i origin/master --exec "sleep 1"
```
before pushing an update. Note that you need to somehow alter a root commit for `sleep 1` to take an effect, otherwise if a commit hasn't changed during the rebase in anyway, it's modification timestamp will not be changed. To modify a commit you can:
- touch some code
- change a comment of a commit
- change base commit (so usually `git fetch` and `git rebase -i origin/master --exec "sleep 1"` is enough on it's own)

pnowojski commented:
Closing, I've squashed the commits and merged it to master.

Thanks for the contribution @ifndef-SleePy

ifndef-SleePy commented:
@pnowojski , wow! The issue has been opened over 4 years. It really surprises me.
Thanks for detailed response. I feel like I have understood Github better and deeper, haha.
